### ISSUE TITLE
Periodic GC job logs are missing after first job run

### ISSUE THREAD

**Issue by dkulchinsky** [goharbor/harbor#14250]
We're running Harbor v2.1.3, jobservice is configured to store job logs in the database.

We have a GC job scheduled to run once a week, the Job executes and finishes successfully, however there is no log stored in the database and we get a "internal server error" (and a 404 in the core log) when trying to retrieve in the UI or via the API.

Job status (indeed shows finished and correct "update" date):
![image](https://user-images.githubusercontent.com/61992380/108247415-a0775780-7120-11eb-861e-653207789b36.png)

Trying to get the log in the UI, we get this:
![image](https://user-images.githubusercontent.com/61992380/108248109-77a39200-7121-11eb-922c-62395bab4671.png)

We are 99.9% (😄) sure we saw the job log on the first scheduled run after the job was configured, but we seem to be hitting this issue on the subsequent job runs.

as a side, I'm also curious what is the expected UX here? since this is a periodic job it has the same ID (105 in our case) every time it runs, so we are only able to access the log of the last run?

`jobservice` log when the job last ran:
```
2021-02-17T00:30:00Z [INFO] [/jobservice/worker/cworker/c_worker.go:76]: Job incoming: {"name":"IMAGE_GC","id":"7d97f3a11a9a6d1eaa4217fe","t":1613521800,"args":null}
2021-02-17T00:30:00Z [INFO] [/jobservice/runner/redis.go:67]: Start to run periodical job execution: 7d97f3a11a9a6d1eaa4217fe@1613521800
2021-02-17T00:30:15Z [INFO] [/jobservice/runner/redis.go:133]: Job 'IMAGE_GC:7d97f3a11a9a6d1eaa4217fe' exit with success
```

`core` log when the job last ran:
```
2021-02-17T00:30:00Z [INFO] [/core/service/notifications/admin/handler.go:98]: received admin job status update event: job-105, job_uuid-7d97f3a11a9a6d1eaa4217fe, status-running, revision-1613521708
2021-02-17T00:30:15Z [INFO] [/core/service/notifications/admin/handler.go:98]: received admin job status update event: job-105, job_uuid-7d97f3a11a9a6d1eaa4217fe, status-finished, revision-1613521708
```

`jobservice` log when trying to retrieve the job log:
```
2021-02-17T15:52:02Z [ERROR] [/jobservice/api/handler.go:288]: Serve http request 'GET /api/v1/jobs/7d97f3a11a9a6d1eaa4217fe/executions?page_number=1&page_size=100' error: 404 {"code":10010,"message":"object is not found","details":"7d97f3a11a9a6d1eaa4217fe"}
```

`core` log when trying to retrieve the job log:
```
2021-02-17T15:52:02Z [ERROR] [/lib/http/error.go:54]: {"errors":[{"code":"INTERNAL_SERVER_ERROR","message":"http error: code 404, message {\"code\":10010,\"message\":\"object is not found\",\"details\":\"7d97f3a11a9a6d1eaa4217fe\"}"}]} 
```

`admin_job` table entry:
```
coredb=# select * from admin_job where id=105;
 id  | job_name | job_kind |                cron_str                 |  status  |         job_uuid         |    creation_time    |        update_time         | deleted |  revision  | status_code |                                                            job_parameters
-----+----------+----------+-----------------------------------------+----------+--------------------------+---------------------+----------------------------+---------+------------+-------------+---------------------------------------------------------------------------------------------------------------------------------------
 105 | IMAGE_GC | Periodic | {"type":"Custom","cron":"0 30 0 * * 3"} | finished | 7d97f3a11a9a6d1eaa4217fe | 2021-02-03 15:14:09 | 2021-02-17 00:30:15.104082 | f       | 1613521708 |           3 | {"delete_untagged":true,"redis_url_reg":"redis+sentinel://rfs-harbor-redis:26379/mymaster/2?idle_timeout_seconds=30","time_window":2}
```

tried to retrieve the log event directly from the job_log table:
```
coredb=# select * from job_log where job_uuid='7d97f3a11a9a6d1eaa4217fe';
 log_id | job_uuid | creation_time | content
--------+----------+---------------+---------
(0 rows)
```

combing through `core` & `jobservice` logs I couldn't find any error or an indication to fail store the job log.

**Comment by bitsf** [comment]
Do you manually run the job again and see if logs exists ?
And check `/var/log/harbor/jobservice.log`, grep `outdated log entries are sweepped`, maybe the logs are deleted

**Comment by dkulchinsky** [comment]
Thanks for your reply @bitsf, if I run GC manually I do see the logs of the manual job.

I also checked the jobservice logs going 7 days back and didn't find anything about `outdated log entries`.

**Comment by steven-zou** [comment]
@dkulchinsky 

Jobservice has a log sweeper mechanism that helps clean the job log with specified # days. By default, it's 1 day.  I guess it's related to this setting.

So far, there is no good way to set that value as it's not exposed outside. We'll plan it in the upcoming release later.

Which installation format are you using? If it is docker-compose way, you can try,

`Update sweeper.duration in <HARBOR_INSTALLATION>/harbor/common/config/jobservice/config.yml`

```
sweeper:
      duration: 10 # 1 #days  #<=======here
      settings: # Customized settings of sweeper
        work_dir: "/var/log/jobs"
```

And then restart the job service container (docker restart harbor-jobservice)or the whole Harbor (by docker-compose -f docker-compose.yml down/ up -d)

**Comment by dkulchinsky** [comment]
Hi @steven-zou, we use a Chart to deploy Harbor in K8s and as I mentioned in my original comment the logs are stored in the database, we also define 14 days for the sweeper duration.

here's the jobservice configuration bit:
```
    job_loggers:
      - name: "DB"
        level: INFO
        sweeper:
          duration: 14 #days
```

I also checked the jobservice logs and didn't see that the sweeper cleaned any logs between the time the job executed and when I checked (which was less than 24 hours)

**Comment by dkulchinsky** [comment]
@steven-zou any thoughts?

**Comment by steven-zou** [comment]
@dkulchinsky 

Sorry for missing the info that you're using the db logger and have already configured the sweeper duration. I'll let the db logger author do a quick check to see whether the db logger does work well or not.

cc @mmpei

**Comment by mmpei** [comment]
i checked the function. it works well. you can check the jobservice log to find a log like 'sweeper xxxxx exit'. it means the sweeper works well.
but the error 'internal server error' is thrown when the log cleaned in db, this maybe a bug. there should be graceful error message.
the scheduled job runs every 7 days, and the sweeper cleans the log recorded before 14 days. it means there are only 2 job_run contains job_log. by the way sweeper only cleans the job log but not the job record. @dkulchinsky

**Comment by dkulchinsky** [comment]
@mmpei, @steven-zou sorry but I'm confused.

As I mentioned, the GC job in question runs weekly on Wed, I'm checking for the log on Thursday (just few hours later) and it's not there, so while the sweeper may be working as expected, the log is still not there. I don't know why and hence the issue.

**Comment by dkulchinsky** [comment]
Trying to work a better reproduction for this, yesterday I disabled the GC schedule (which removed job 105 mentioned above), then re-enabled the schedule so it created a new job (id=121) and it ran last night and produced a log entry which we can retrieve in the UI
![image](https://user-images.githubusercontent.com/61992380/113152389-1385ea80-9204-11eb-8d25-a911e360e7ec.png)

![image](https://user-images.githubusercontent.com/61992380/113152464-28627e00-9204-11eb-8bf1-9ad84c5c079e.png)

so thus far this works as we described - the first scheduled execution produces a log.

```
coredb=# select * from admin_job where id=121;
 id  | job_name | job_kind |                cron_str                 |  status  |         job_uuid         |    creation_time    |        update_time         | deleted |  revision  | status_code |                                                            job_parameters
-----+----------+----------+-----------------------------------------+----------+--------------------------+---------------------+----------------------------+---------+------------+-------------+---------------------------------------------------------------------------------------------------------------------------------------
 121 | IMAGE_GC | Periodic | {"type":"Custom","cron":"0 30 0 * * 3"} | finished | 5a4ea58f6b40f7011298a925 | 2021-03-30 02:32:36 | 2021-03-31 11:50:36.469947 | f       | 1617150513 |           3 | {"delete_untagged":true,"redis_url_reg":"redis+sentinel://rfs-harbor-redis:26379/mymaster/2?idle_timeout_seconds=30","time_window":2}
```

and querying the `job_log` table with above `job_uuid` produced the log as well:
```
coredb=# select * from job_log where job_uuid like '%5a4ea58f6b40f7011298a925%';
 log_id |              job_uuid               |       creation_time        |                                                                                                                                                                                                          content
--------+-------------------------------------+----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 148473 | 5a4ea58f6b40f7011298a925@1617150600 | 2021-03-31 11:50:36.350773 | 2021-03-31T00:30:00Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:144]: Garbage Collection parameters: [delete_untagged: true, dry_run: false, time_window: 2]                                                                                                                                                                                                                                                   +
        |                                     |                            | 2021-03-31T00:30:00Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:154]: start to run gc in job.                                                                                                                                                                                                                                                                                                                  +
        |                                     |                            | 2021-03-31T00:30:00Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:380]: start to delete untagged artifact.
```

I'm trying to understand the logic here, since the `id` in `admin_job` table doesn't change over time, is the `job_uuid` overwritten every time the job is scheduled? in any case, will need to wait until the next time GC runs and see if I can retrieve the job log.

**Comment by dkulchinsky** [comment]
A week has passed and the GC job was executed, it finished successfully however trying to get the log now in the UI produces the `internal server error` error (which btw should be changed to something like `GC job log not found`).

Looking at the `admin_job` table now, I see that the job has finished and update_time set accordingly.
```
coredb=# select * from admin_job where id=121;
 id  | job_name | job_kind |                cron_str                 |  status  |         job_uuid         |    creation_time    |        update_time         | deleted |  revision  | status_code |                                                            job_parameters
-----+----------+----------+-----------------------------------------+----------+--------------------------+---------------------+----------------------------+---------+------------+-------------+---------------------------------------------------------------------------------------------------------------------------------------
 121 | IMAGE_GC | Periodic | {"type":"Custom","cron":"0 30 0 * * 3"} | finished | 5a4ea58f6b40f7011298a925 | 2021-03-30 02:32:36 | 2021-04-07 06:25:01.634651 | f       | 1617755362 |           3 | {"delete_untagged":true,"redis_url_reg":"redis+sentinel://rfs-harbor-redis:26379/mymaster/2?idle_timeout_seconds=30","time_window":2}
```

looking at the `job_log` table, I see the new entry == 165972 (I also see the previous one from last week):
```
coredb=# select log_id,job_uuid,creation_time from job_log where job_uuid like '%5a4ea58f6b40f7011298a925%';
 log_id |              job_uuid               |       creation_time
--------+-------------------------------------+----------------------------
 148473 | 5a4ea58f6b40f7011298a925@1617150600 | 2021-03-31 11:50:36.350773
 165972 | 5a4ea58f6b40f7011298a925@1617755400 | 2021-04-07 06:25:01.568297
(2 rows)
```

however, core service returns this when I try to get the log in the UI:

```
[harbor-core-7687dcbfd4-mnsqw core] 2021-04-08T01:25:12Z [DEBUG] [/server/middleware/log/log.go:30]: attach request id e8d126bc9c549eb3ac6af5195187eac0 to the logger for the request GET /api/v2.0/system/gc/121/log
[harbor-core-7687dcbfd4-mnsqw core] 2021-04-08T01:25:12Z [DEBUG] [/server/middleware/artifactinfo/artifact_info.go:52]: In artifact info middleware, url: /api/v2.0/system/gc/121/log
[harbor-core-7687dcbfd4-mnsqw core] 2021-04-08T01:25:12Z [DEBUG] [/common/utils/oidc/helper.go:194]: Raw ID token for verification:
[harbor-core-7687dcbfd4-mnsqw core] 2021-04-08T01:25:12Z [WARNING] [/server/middleware/security/idtoken.go:45][requestID="e8d126bc9c549eb3ac6af5195187eac0"]: failed to verify token: oidc: malformed jwt: square/go-jose: compact JWS format must have three parts
[harbor-core-7687dcbfd4-mnsqw core] 2021-04-08T01:25:12Z [DEBUG] [/server/middleware/security/session.go:47][requestID="e8d126bc9c549eb3ac6af5195187eac0"]: a session security context generated for request GET /api/v2.0/system/gc/121/log
[harbor-core-7687dcbfd4-mnsqw core] 2021-04-08T01:25:12Z [ERROR] [/lib/http/error.go:54]: {"errors":[{"code":"INTERNAL_SERVER_ERROR","message":"http error: code 404, message {\"code\":10010,\"message\":\"object is not found\",\"details\":\"5a4ea58f6b40f7011298a925\"}"}]}
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/lib/http/error.go:82, github.com/goharbor/harbor/src/lib/http.apiError
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/lib/http/error.go:51, github.com/goharbor/harbor/src/lib/http.SendError
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/common/api/base.go:67, github.com/goharbor/harbor/src/common/api.(*BaseAPI).RenderError
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/common/api/base.go:228, github.com/goharbor/harbor/src/common/api.(*BaseAPI).SendInternalServerError
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/core/api/admin_job.go:184, github.com/goharbor/harbor/src/core/api.(*AJAPI).getLog
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/core/api/reg_gc.go:146, github.com/goharbor/harbor/src/core/api.(*GCAPI).GetLog
[harbor-core-7687dcbfd4-mnsqw core] /usr/local/go/src/reflect/value.go:460, reflect.Value.call
[harbor-core-7687dcbfd4-mnsqw core] /usr/local/go/src/reflect/value.go:321, reflect.Value.Call
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/vendor/github.com/astaxie/beego/router.go:853, github.com/astaxie/beego.(*ControllerRegister).ServeHTTP
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/server/middleware/transaction/transaction.go:62, github.com/goharbor/harbor/src/server/middleware/transaction.Middleware.func1.1
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/lib/orm/orm.go:76, github.com/goharbor/harbor/src/lib/orm.WithTransaction.func1
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/server/middleware/transaction/transaction.go:71, github.com/goharbor/harbor/src/server/middleware/transaction.Middleware.func1
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/server/middleware/middleware.go:57, github.com/goharbor/harbor/src/server/middleware.New.func1.1
[harbor-core-7687dcbfd4-mnsqw core] /usr/local/go/src/net/http/server.go:2041, net/http.HandlerFunc.ServeHTTP
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/server/middleware/middleware.go:52, github.com/goharbor/harbor/src/server/middleware.New.func1.1
[harbor-core-7687dcbfd4-mnsqw core] /usr/local/go/src/net/http/server.go:2041, net/http.HandlerFunc.ServeHTTP
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/server/middleware/security/security.go:63, github.com/goharbor/harbor/src/server/middleware/security.Middleware.func1.1
[harbor-core-7687dcbfd4-mnsqw core] /usr/local/go/src/net/http/server.go:2041, net/http.HandlerFunc.ServeHTTP
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/server/middleware/artifactinfo/artifact_info.go:55, github.com/goharbor/harbor/src/server/middleware/artifactinfo.Middleware.func1.1
[harbor-core-7687dcbfd4-mnsqw core] /usr/local/go/src/net/http/server.go:2041, net/http.HandlerFunc.ServeHTTP
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/server/middleware/notification/notification.go:31, github.com/goharbor/harbor/src/server/middleware/notification.Middleware.func1
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/server/middleware/middleware.go:57, github.com/goharbor/harbor/src/server/middleware.New.func1.1
[harbor-core-7687dcbfd4-mnsqw core] /usr/local/go/src/net/http/server.go:2041, net/http.HandlerFunc.ServeHTTP
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/server/middleware/orm/orm.go:53, github.com/goharbor/harbor/src/server/middleware/orm.MiddlewareWithConfig.func1
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/server/middleware/middleware.go:57, github.com/goharbor/harbor/src/server/middleware.New.func1.1
[harbor-core-7687dcbfd4-mnsqw core] /usr/local/go/src/net/http/server.go:2041, net/http.HandlerFunc.ServeHTTP
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/server/middleware/csrf/csrf.go:48, github.com/goharbor/harbor/src/server/middleware/csrf.attach.func1
[harbor-core-7687dcbfd4-mnsqw core] /usr/local/go/src/net/http/server.go:2041, net/http.HandlerFunc.ServeHTTP
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/vendor/github.com/gorilla/csrf/csrf.go:297, github.com/gorilla/csrf.(*csrf).ServeHTTP
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/server/middleware/csrf/csrf.go:68, github.com/goharbor/harbor/src/server/middleware/csrf.Middleware.func2
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/server/middleware/middleware.go:57, github.com/goharbor/harbor/src/server/middleware.New.func1.1
[harbor-core-7687dcbfd4-mnsqw core] /usr/local/go/src/net/http/server.go:2041, net/http.HandlerFunc.ServeHTTP
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/server/middleware/session/session.go:34, github.com/goharbor/harbor/src/server/middleware/session.Middleware.func1.1
[harbor-core-7687dcbfd4-mnsqw core] /usr/local/go/src/net/http/server.go:2041, net/http.HandlerFunc.ServeHTTP
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/server/middleware/log/log.go:33, github.com/goharbor/harbor/src/server/middleware/log.Middleware.func1
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/server/middleware/middleware.go:57, github.com/goharbor/harbor/src/server/middleware.New.func1.1
[harbor-core-7687dcbfd4-mnsqw core] /usr/local/go/src/net/http/server.go:2041, net/http.HandlerFunc.ServeHTTP
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/server/middleware/requestid/requestid.go:37, github.com/goharbor/harbor/src/server/middleware/requestid.Middleware.func1
[harbor-core-7687dcbfd4-mnsqw core] /harbor/src/server/middleware/middleware.go:57, github.com/goharbor/harbor/src/server/middleware.New.func1.1
[harbor-core-7687dcbfd4-mnsqw core] /usr/local/go/src/net/http/server.go:2041, net/http.HandlerFunc.ServeHTTP
[harbor-core-7687dcbfd4-mnsqw core] /usr/local/go/src/net/http/server.go:2836, net/http.serverHandler.ServeHTTP
[harbor-core-7687dcbfd4-mnsqw core] /usr/local/go/src/net/http/server.go:1924, net/http.(*conn).serve
```

by all accounts, this seems like a bug and it's not a log sweeper in action because both the latest and previous job logs are still there in the database as I shown above.

@mmpei @steven-zou @xaleeks I think I provided enough debug information to show that something is broken with retrieval of the GC logs from the second scheduled run and onward and it is not a sweeper issue (which is set to 14 days, so shouldn't even be a concern here anyway).

**Comment by dkulchinsky** [comment]
From what I can tell, the error bubbles up from here:
https://github.com/goharbor/harbor/blob/e2426603fd0be8d14138c113d10b278de5f4876c/src/core/api/admin_job.go#L182-L186

which seems to be failing to retrieve the executions, is there an implementation bug there when there are multiple job execution for the same periodic GC job?

**Comment by dkulchinsky** [comment]
@steven-zou @mmpei @xaleeks I'd really appreciate your help on this 🙏🏼 we just had a scheduled GC job fail and I can't see the logs to understand why, it's really frustrating... more than happy to provide any additional information if needed.

```
coredb=# select * from admin_job where id=125;
 id  | job_name | job_kind |                cron_str                 | status |         job_uuid         |    creation_time    |        update_time         | deleted |  revision  | status_code |                                                            job_parameters
-----+----------+----------+-----------------------------------------+--------+--------------------------+---------------------+----------------------------+---------+------------+-------------+---------------------------------------------------------------------------------------------------------------------------------------
 125 | IMAGE_GC | Periodic | {"type":"Custom","cron":"0 30 0 * * 3"} | error  | 2321227a3453c7311477573c | 2021-04-19 14:27:16 | 2021-04-20 14:27:25.300521 | f       | 1618842436 |           3 | {"delete_untagged":true,"redis_url_reg":"redis+sentinel://rfs-harbor-redis:26379/mymaster/2?idle_timeout_seconds=30","time_window":2}
```

I can't even see the job log in `job_log` table:
```
coredb=# select log_id,job_uuid,creation_time from job_log where job_uuid like '%2321227a3453c7311477573c%';
 log_id | job_uuid | creation_time
--------+----------+---------------
(0 rows)
```

**Comment by dkulchinsky** [comment]
ping 🙏🏼

**Comment by dkulchinsky** [comment]
We've recently upgraded to v2.1.5 and the issue persists, can someone please take a look?

@steven-zou @bitsf @mmpei

**Comment by dkulchinsky** [comment]
Trying to investigate this further, the `utils_core.GetJobServiceClient().GetExecutions(job.UUID)` is querying the `jobservice` API to retrieve the the executions here:
https://github.com/goharbor/harbor/blob/919c9fd2c63b409edd64f4988956379e563136c2/src/common/job/client.go#L163-L164

I executed the same query with `curl` (port forwarding to a `jobservice` Pod):
```
❯ curl "localhost:8080/api/v1/jobs/2321227a3453c7311477573c/executions?page_number=1&page_size=100"  -H "authorization: Harbor-Secret [REDACTED]"
{"code":10010,"message":"object is not found","details":"2321227a3453c7311477573c"}
```

when running the query above, I see this error in `jobservice`:
```
[harbor-jobservice-7d7874fc97-6xvlr jobservice] 2021-05-31T21:11:51Z [ERROR] [/jobservice/api/handler.go:288]: Serve http request 'GET /api/v1/jobs/2321227a3453c7311477573c/executions?page_number=1&page_size=100' error: 404 {"code":10010,"message":"object is not found","details":"2321227a3453c7311477573c"}
```

which is the same error when trying to retrieve the log from the UI.

**Comment by steven-zou** [comment]
@dkulchinsky 

Sorry for the late response and thanks a lot for providing the related info. We'll take a look at it and see if we can find the concrete root cause and fix it ASAP.

@wy65701436 @ywk253100 let's have an offline discussion later.

**Comment by steven-zou** [comment]
> @steven-zou @mmpei @xaleeks I'd really appreciate your help on this 🙏🏼 we just had a scheduled GC job fail and I can't see the logs to understand why, it's really frustrating... more than happy to provide any additional information if needed.
> 
> ```
> coredb=# select * from admin_job where id=125;
>  id  | job_name | job_kind |                cron_str                 | status |         job_uuid         |    creation_time    |        update_time         | deleted |  revision  | status_code |                                                            job_parameters
> -----+----------+----------+-----------------------------------------+--------+--------------------------+---------------------+----------------------------+---------+------------+-------------+---------------------------------------------------------------------------------------------------------------------------------------
>  125 | IMAGE_GC | Periodic | {"type":"Custom","cron":"0 30 0 * * 3"} | error  | 2321227a3453c7311477573c | 2021-04-19 14:27:16 | 2021-04-20 14:27:25.300521 | f       | 1618842436 |           3 | {"delete_untagged":true,"redis_url_reg":"redis+sentinel://rfs-harbor-redis:26379/mymaster/2?idle_timeout_seconds=30","time_window":2}
> ```
> 
> I can't even see the job log in `job_log` table:
> 
> ```
> coredb=# select log_id,job_uuid,creation_time from job_log where job_uuid like '%2321227a3453c7311477573c%';
>  log_id | job_uuid | creation_time
> --------+----------+---------------
> (0 rows)
> ```

@dkulchinsky 

For this case, you need to check the `core.log` and `jobservice.log` because the job may not be successfully launched/executed. In that case, no job log might be generated.

**Comment by steven-zou** [comment]
@dkulchinsky 

We have located the root cause of this issue. It is because the job stats data kept in the Redis has an expiration time. The expiration time is hardcoded and set to be 1 day. That means 1 day later the data will be gone and no chance to retrieve it. However, the core service needs to query it when getting the log. As it has been gone and the issue is raised.

The expiration time is by design and the purpose of it is to reduce the storage size of the Redis as there might be large-scale job data generated in a short time.

In 2.2.x (now the latest patch is 2.2.2), we have used a different way to launch GC jobs.  There will be no issues in 2.2.x. So, a possible way is upgrading your Harbor to 2.2.2 with harbor helm v1.6.2.

As it's very hard to fix it in the 2.1 branch, so we don't prefer to introduce any fix into 2.1.x (the expiration is by design and it has been fixed in 2.2.x). Maybe an upgrade is a good option now.

**Comment by dkulchinsky** [comment]
Hey @steven-zou, thanks for the detailed analysis and root cause.

Few notes on our end:

1. We're in the process of certifying v2.2.2 in our test environment, however after upgrading to v2.2.2 from v2.1.5 we're no longer able to sign our images https://github.com/goharbor/harbor/issues/14932#issuecomment-862927908 and looks like few other users are reporting a similar issue, so for now we're blocked from upgrading our production Harbor until this is resolved.

2. Is it possible to at least fix the returned message? to display something like "Job log not found" instead of "Internal Server Error"? would be a better UX at least.

**Comment by dkulchinsky** [comment]
Going to close this now, issue has been resolved in v2.2 and onward, thanks!
